x-pack/plugins/security_solution/public/timelines/components/use_timeline_save_prompt/index.ts

```diff
@@ -0,0 +1,126 @@
+import { useKibana } from '../../../common/lib/kibana';
+import { useDeepEqualSelector } from '../../../common/hooks/use_selector';
+import { getTimelineShowStatusByIdSelector } from '../flyout/selectors';
+import { getLinksWithHiddenTimeline } from '../../../common/links';
+
+// Issue with history.block
+// https://github.com/elastic/kibana/issues/132597
+
+export const useTimelineSavePrompt = (
+  timelineId: TimelineId,
+  onAppLeave: (handler: AppLeaveHandler) => void
+) => {
+  const dispatch = useDispatch();
+
+  const linksWithoutTimelines = getLinksWithHiddenTimeline();
+
+  const pathsWithoutTimelines = linksWithoutTimelines.map((link) => `/app/security${link.path}`);
+
+  const { overlays, application } = useKibana().services;
+  const history = useHistory();
+
+  const prompt = i18n.translate('xpack.securitySolution.timeline.unsavedWorkMessage', {
+    defaultMessage: 'Leave Timeline with unsaved work?',
+  });
+
+  const title = i18n.translate('xpack.securitySolution.timeline.unsavedWorkTitle', {
+    defaultMessage: 'Unsaved changes',
+  });
+
+  const getTimelineShowStatus = useMemo(() => getTimelineShowStatusByIdSelector(), []);
+  const { status: timelineStatus, updated } = useDeepEqualSelector((state) =>
+    getTimelineShowStatus(state, timelineId)
+  );
+
+  const showSaveTimelineModal = useCallback(() => {
+    dispatch(timelineActions.showTimeline({ id: timelineId, show: true }));
+    dispatch(
+      timelineActions.setActiveTabTimeline({
+        id: timelineId,
+        activeTab: TimelineTabs.query,
+      })
+    );
+    dispatch(
+      timelineActions.toggleModalSaveTimeline({
+        id: timelineId,
+        showModalSaveTimeline: true,
+      })
+    );
+  }, [dispatch, timelineId]);
+
+  useEffect(() => {
+    const unblock = history.block((location) => {
+      async function confirmSaveTimeline() {
+        const confirmRes = await overlays?.openConfirm(prompt, {
+          title,
+          'data-test-subj': 'appLeaveConfirmModal',
+        });
+
+        if (confirmRes) {
+          unblock();
+
+          application.navigateToUrl(location.pathname + location.hash + location.search, {
+            state: location.state,
+          });
+        } else {
+          showSaveTimelineModal();
+        }
+      }
+
+      if (
+        pathsWithoutTimelines.includes(location.pathname) &&
+        timelineStatus === TimelineStatus.draft &&
+        updated != null
+      ) {
+        confirmSaveTimeline();
+      } else {
+        return;
+      }
+      return false;
+    });
+
+    return () => {
+      unblock();
+    };
+  }, [
+    history,
+    application,
+    overlays,
+    showSaveTimelineModal,
+    prompt,
+    title,
+    timelineStatus,
+    updated,
+    pathsWithoutTimelines,
+  ]);
+
+  useEffect(() => {
+    onAppLeave((actions, nextAppId) => {
+      // Confirm when the user has made any changes to a timeline
+      if (
+        !(nextAppId ?? '').includes('securitySolution') &&
```

Comment:
It is actually APP_ID. I am not sure but I think it is because `onAppLeave` hook is owned by kibana core and we are registering out app using `APP_ID` and not `APP_UI_ID`. `APP_UI_ID` is being used in the telemetry though.

https://github.com/elastic/kibana/blob/0fbbd4f18ae78e4346366152192c4b2e0455ba24/x-pack/plugins/security_solution/public/plugin.tsx#L100-L123

Reply:
The Security plugin is registered using `APP_UI_ID`:
https://github.com/elastic/kibana/blob/6ae4764db6ed54ac8ffc690ac4f1c4c81ad176e1/x-pack/plugins/security_solution/public/plugin.tsx#L152-L153

However, I have been trying to print this guy to confirm it, but there's no way `nextAppId` to be our ID, it does not trigger when navigating within Security, only when navigating to another app, or reloading (takes undefined), so probably this ID check it not even needed 😂 

I won't be picky here, do whatever you think is best 👍

Reply:
You are right..  we raised the modal when `nextAppId` is NOT security Solution i.e. when user is navigating out of security solution and they have an unsaved timeline.  It is an existing piece of code which I refactored into a single hook. You can check the cypress test which demonstrate this use case.